<?xml version="1.0"?>
<launch>
  <!-- 1. Виртуальное время -->
  <param name="use_sim_time" value="true"/>

  <!-- 2. MAVROS SITL -->
  <include file="$(find mavros)/launch/apm.launch">
    <arg name="fcu_url" value="udp://:14550@127.0.0.1:14557"/>
  </include>

  <!-- 3. Загрузка URDF -->
  <param name="robot_description" command="$(find xacro)/xacro '$(find basic_nav)/urdf/drone_urdf.urdf'"/>
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="state_publisher" output="screen">
    <param name="publish_frequency" value="20.0"/>
  </node>

  <!-- 4. Статический трансформ map -> odom -->
  <node pkg="tf" type="static_transform_publisher" name="static_map_to_odom"
        args="0 0 0 0 0 0 map odom 100" />

  <!-- 5. Статический трансформ odom -> base_link (временно для инициализации) -->
  <node pkg="tf" type="static_transform_publisher" name="initial_pose_tf"
        args="0.0 0.0 0 0 0 0 odom base_link 100" />
        
  <!-- 6. Статический трансформ base_link -> lidar_link -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_lidar"
        args="0 0 0.2 0 0 0 base_link lidar_link 100" />

  <!-- 7. Контроллер прямого управления дроном -->
  <node name="mavros_direct_control" pkg="basic_nav" type="mavros_direct_control.py" output="screen">
    <param name="takeoff_height" value="2.0"/>
    <param name="arm_timeout" value="60.0"/>
    <param name="command_rate" value="20.0"/>
  </node>

  <!-- 8. Координатный синхронизатор -->
  <node name="coordinate_sync" pkg="basic_nav" type="coordinate_sync.py" output="screen">
    <param name="use_mavros" value="true"/>
    <param name="use_odom" value="true"/>
    <param name="publish_tf" value="true"/>
    <param name="sync_rate" value="30.0"/>
    <param name="force_map_origin" value="true"/>
  </node>

  <!-- 9. Визуализация в RViz -->
  <node name="visualization_helper" pkg="basic_nav" type="visualization_helper.py" output="screen">
    <param name="marker_size" value="0.5"/>
    <param name="update_rate" value="10.0"/>
  </node>

  <!-- 10. Создание пустой карты для визуализации -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(find basic_nav)/maps/empty_map.yaml">
    <param name="frame_id" value="map"/>
  </node>

  <!-- 11. RViz для визуализации -->
  <node pkg="rviz" type="rviz" name="rviz" 
        args="-d $(find basic_nav)/rviz/drone_direct_control.rviz" required="false"/>
</launch> 